### DISCUSSION TITLE
Custom AsyncMethodBuilder method overloads

### DISCUSSION THREAD

**Discussion by timcassell** [dotnet/runtime#61574] [General]
I tried adding method overloads for my custom builder to add a feature for my library in async functions, using the awaiter type.

```cs
public struct PromiseMethodBuilder<T>
{
    ...

    [SecuritySafeCritical]
    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
        where TAwaiter : ICriticalNotifyCompletion
        where TStateMachine : IAsyncStateMachine
    {
        Console.WriteLine("AwaitUnsafeOnCompleted default");
        ...
    }

    [SecuritySafeCritical]
    public void AwaitUnsafeOnCompleted<TStateMachine>(ref PromiseAwaiterVoid awaiter, ref TStateMachine stateMachine)
        where TStateMachine : IAsyncStateMachine
    {
        Console.WriteLine("AwaitUnsafeOnCompleted PromiseAwaiterVoid");
        ...
    }

    [SecuritySafeCritical]
    public void AwaitUnsafeOnCompleted<T, TStateMachine>(ref PromiseAwaiter<T> awaiter, ref TStateMachine stateMachine)
        where TStateMachine : IAsyncStateMachine
    {
        Console.WriteLine("AwaitUnsafeOnCompleted PromiseAwaiter<T>");
        ...
    }
    
    ...
}
```

Using the builder with those overloads directly works as I expect:

```cs
var deferred = Promise.NewDeferred<int>();
PromiseAwaiter<int> awaiter = deferred.Promise.GetAwaiter();
IAsyncStateMachine stateMachine = null;
new PromiseMethodBuilder<int>().AwaitUnsafeOnCompleted(ref awaiter, ref stateMachine); // Prints "AwaitUnsafeOnCompleted PromiseAwaiter<T>"
```

But when I tested it with an actual `async Promise` function, it printed `AwaitUnsafeOnCompleted default`.

```cs
var deferred = Promise.NewDeferred<int>();

async Promise<int> Func()
{
    return await deferred.Promise; // Prints "AwaitUnsafeOnCompleted default"
}
```

So the compiler chose to use the default overload despite the existence of the better overload candidate that the direct call uses.

Is this possible to work this way? Does this need to be a new feature request to allow this?

I'm not sure if this should be in runtime or csharplang.

@stephentoub 

[Edit] I understand _why_ this is the case (the compiler always includes the `<TAwaiter, TStateMachine>` type arguments to the method call explicitly), but is it possible to override that for specific types?

**Comment by stephentoub** [comment]
You would need an addition to the language/ compiler.  But why do you need overloads rather than just dynamically type testing the TAwaiter? The latter is what the built-in builders do for special-casing the awaiters they know about, no overloads required.

**Reply by timcassell** [reply]
Because I don't have access to the `<T>` type when the awaiter is `PromiseAwaiter<T>`. How can I type-check and cast that?

**Reply by stephentoub** [reply]
The same same way we do for `TaskAwaiter<T>`, an internal interface.

**Reply by timcassell** [reply]
Oh I see, you rely on `TaskAwaiter` and `TaskAwaiter<T>` to have the same structure. It looks like that's possible because they just wrap a reference. But what if I can't rely on that for my awaiters? My `Promise<T>` are structs that contain a reference field, some other small fields (`short`, `int`), and a `T` field with automatic layout, so their layout could be different depending on what `T` is.

**Reply by stephentoub** [reply]
> But what if I can't rely on that for my awaiters? 

Add methods to the interface to interact with them however you need to.

**Reply by timcassell** [reply]
But wouldn't that require boxing? I want to avoid boxing.

**Reply by stephentoub** [reply]
No. We do exactly that.  It's a constrained call.  There's no boxing. 
https://github.com/dotnet/runtime/blob/a3b186b9904844b8001d4bb5d7474f2591ee96d9/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs#L119

**Reply by timcassell** [reply]
Huh, interesting. I was under the impression that that produced a boxing operation to be able to call the interface method. Has constrained calling always been part of the runtime? A quick google search doesn't give me much information, do you have a resource you can link me to so I can educate myself some more on the topic?

**Reply by timcassell** [reply]
Also, would you know if constrained calls are optimized like that in the Unity engine? Or does it box like I assumed before? Or does that depend on Mono vs IL2CPP?

**Reply by timcassell** [reply]
I tested this out in sharplab.io

```cs
public void M<T>(ref T awaiter)
{
    if (null != default(T) && awaiter is IMyInterface)
    {
        ((IMyInterface)awaiter).MyMethod();
    }
}
```

The output IL is

```cs
.method public hidebysig 
        instance void M<T> (
            !!T& awaiter
        ) cil managed 
    {
        // Method begins at RVA 0x2090
        // Code size 56 (0x38)
        .maxstack 1
        .locals init (
            [0] !!T
        )

        IL_0000: ldloca.s 0
        IL_0002: initobj !!T
        IL_0008: ldloc.0
        IL_0009: box !!T
        IL_000e: brfalse.s IL_0037

        IL_0010: ldarg.1
        IL_0011: ldobj !!T
        IL_0016: box !!T
        IL_001b: isinst IMyInterface
        IL_0020: brfalse.s IL_0037

        IL_0022: ldarg.1
        IL_0023: ldobj !!T
        IL_0028: box !!T
        IL_002d: castclass IMyInterface
        IL_0032: callvirt instance void IMyInterface::MyMethod()

        IL_0037: ret
    } // end of method C::M
```

There's a clear box instruction there (actually 3!) and a callvirt, not the constrained call I expected. Is sharplab.io generating different IL than Roslyn?

**Reply by stephentoub** [reply]
> There's a clear box instruction there (actually 3!) and a callvirt, not the constrained 

They're all eliminated by the JIT.

**Reply by timcassell** [reply]
> > There's a clear box instruction there (actually 3!) and a callvirt, not the constrained
> 
> They're all eliminated by the JIT.

Ah ok, so I can't rely on that optimization in Unity, then. Guess I'll have to work around this another way.

**Reply by stephentoub** [reply]
> Ah ok, so I can't rely on that optimization in Unity

You would need to confirm with the compilers used there.

**Reply by DaZombieKiller** [reply]
@timcassell You can use https://github.com/sschoener/unity-asm-explorer-package to view the generated assembly in Unity, though I'm not sure if it still works with the latest versions of the engine (I have not personally tested). Of course, this doesn't apply to IL2CPP, only the Mono JIT.

**Reply by timcassell** [reply]
[Edit] Sorry, I misread.

**Reply by timcassell** [reply]
If I used a sequential layout instead of auto, could I rely on that for an unsafe cast? Since I don't actually care what the `T` is for this purpose.

```cs
public readonly struct Promise
{
    internal readonly Promise<Internal.VoidResult> _target;
}

[StructLayout(LayoutKind.Sequential)]
public readonly struct Promise<T>
{
    internal readonly Internal.PromiseRef _ref;
    internal readonly int _depth;
    internal readonly short _id;
    internal readonly T _result;
}

public readonly struct PromiseAwaiter : IPromiseAwaiter
{
    internal readonly PromiseAwaiter<Internal.VoidResult> _target;
}

public readonly struct PromiseAwaiter<T> : IPromiseAwaiter
{
    internal readonly Promise<T> _target;
}
```

...

```cs
Unsafe.As<TAwaiter, PromiseAwaiter>(ref awaiter).AwaitUnsafeOnCompletedInternal(ref this);
```

**Reply by timcassell** [reply]
For anyone also wondering, I just profiled builds in Unity, and it looks like neither the Mono nor IL2CPP runtimes optimize away the `((IMyInterface)awaiter).MyMethod();` or `awaiter is IMyInterface` boxes, but they do both optimize away the `null != default(T)` box.

**Reply by timcassell** [reply]
This is my workaround to get it to work without boxing in Unity:

```cs
internal abstract class AwaitOverrider<T> where T : INotifyCompletion
{
    internal static AwaitOverrider<T> awaitOverrider;

    internal abstract void AwaitOnCompletedInternal(ref T awaiter, Internal.AsyncPromiseRef asyncPromiseRef);
}

internal sealed class AwaitOverriderImpl<TAwaiter> : AwaitOverrider<TAwaiter> where TAwaiter : struct, ICriticalNotifyCompletion, IPromiseAwaiter
{
    internal static void Create()
    {
        awaitOverrider = new AwaitOverriderImpl<TAwaiter>();
    }

    internal override void AwaitOnCompletedInternal(ref TAwaiter awaiter, Internal.AsyncPromiseRef asyncPromiseRef)
    {
        awaiter.AwaitOnCompletedInternal(asyncPromiseRef);
    }
}

internal interface IPromiseAwaiter
{
    void AwaitOnCompletedInternal(Internal.AsyncPromiseRef asyncPromiseRef);
}

public struct PromiseAwaiter<T> : ICriticalNotifyCompletion, IPromiseAwaiter
{
    static PromiseAwaiter()
    {
        AwaitOverriderImpl<PromiseAwaiter<T>>.Create();
    }

    void IPromiseAwaiter.AwaitOnCompletedInternal(Internal.AsyncPromiseRef asyncPromiseRef)
    {
        // Internal implementation...
    }
}
```

```cs
if (null != default(TAwaiter) && typeof(IPromiseAwaiter).IsAssignableFrom(typeof(TAwaiter)))
{
    AwaitOverrider<TAwaiter>.awaitOverrider.AwaitOnCompletedInternal(ref awaiter, _promiseRef);
}
else
{
    // Normal path...
}
```

This could also be optimized more with a function pointer instead of an object instance, but I'm still supporting older Unity versions with C#7. And the `IsAssignableFrom` could be changed to a null check, but I haven't benchmarked that to see if it would be faster.
